pkg/controller/direct/alloydb/cluster_controller.go

```diff
@@ -17,339 +17,644 @@ package alloydb
+		if err != nil {
+			return err
+		}
+		obj.Spec.RestoreContinuousBackupSource.ClusterRef.External = external
+	}
 
-	log.V(0).Info("creating cluster", "cluster", cluster)
-	var op *api.Operation
-	var err error
-	// Default ClusterType to be PRIMARY
-	if cluster.ClusterType == "SECONDARY" {
-		op, err = a.client.Projects.Locations.Clusters.Createsecondary(parent, cluster).ClusterId(a.resourceID).Context(ctx).Do()
-	} else {
-		op, err = a.client.Projects.Locations.Clusters.Create(parent, cluster).ClusterId(a.resourceID).Context(ctx).Do()
+	if obj.Spec.SecondaryConfig != nil && obj.Spec.SecondaryConfig.PrimaryClusterNameRef != nil {
+		external, err := obj.Spec.SecondaryConfig.PrimaryClusterNameRef.NormalizedExternal(ctx, a.reader, obj.Namespace)
+		if err != nil {
+			return err
+		}
+		obj.Spec.SecondaryConfig.PrimaryClusterNameRef.External = external
 	}
-	if err != nil {
-		return fmt.Errorf("creating cluster: %w", err)
+
+	return nil
+}
+
+// TODO: Scenario test case: ContinuousBackupConfig.Enabled unset.
+func (a *ClusterAdapter) resolveKRMDefaultsForCreate() {
+	obj := a.desired
+	if obj.Spec.ClusterType == nil || direct.ValueOf(obj.Spec.ClusterType) == "" {
+		obj.Spec.ClusterType = direct.LazyPtr("PRIMARY")
 	}
+	if obj.Spec.ContinuousBackupConfig != nil && obj.Spec.ContinuousBackupConfig.Enabled == nil {
+		obj.Spec.ContinuousBackupConfig.Enabled = direct.PtrTo(true)
+	}
+	if obj.Spec.DeletionPolicy == nil || direct.ValueOf(obj.Spec.DeletionPolicy) == "" {
+		obj.Spec.DeletionPolicy = direct.LazyPtr("DEFAULT")
+	}
+}
 
-	if err := a.waitForOp(ctx, op); err != nil {
-		return fmt.Errorf("waiting for cluster create %s: %w", a.fullyQualifiedName(), err)
+func (a *ClusterAdapter) resolveKRMDefaultsForUpdate() {
+	a.resolveKRMDefaultsForCreate()
+	obj := a.desired
+	// This is needed for only update because the returned actual state has both
+	// fields set to the same value.
+	if obj.Spec.NetworkRef == nil && obj.Spec.NetworkConfig != nil && obj.Spec.NetworkConfig.NetworkRef != nil {
+		obj.Spec.NetworkRef = &refs.ComputeNetworkRef{
+			External: obj.Spec.NetworkConfig.NetworkRef.External,
+		}
+	} else if (obj.Spec.NetworkConfig == nil || obj.Spec.NetworkConfig.NetworkRef == nil) && obj.Spec.NetworkRef != nil {
+		if obj.Spec.NetworkConfig == nil {
+			obj.Spec.NetworkConfig = &krm.Cluster_NetworkConfig{}
+		}
+		obj.Spec.NetworkConfig.NetworkRef = &refs.ComputeNetworkRef{
+			External: obj.Spec.NetworkRef.External,
+		}
 	}
+}
 
-	created, err := a.client.Projects.Locations.Clusters.Get(a.fullyQualifiedName()).Context(ctx).Do()
-	if err != nil {
-		return fmt.Errorf("getting created cluster: %w", err)
+// TODO: Scenario test case: Update initialUser.password from `value` to `valueFrom` and vise versa.
+func (a *ClusterAdapter) resolveInitialUserField(ctx context.Context) error {
+	obj := a.desired
+	if obj.Spec.InitialUser != nil && obj.Spec.InitialUser.Password != nil {
+		value := obj.Spec.InitialUser.Password.Value
```

Comment:
Consider packaging this logic into a helper function to resolve the password from "refsv1beta1secret.Legacy"?

Reply:
Good point. Will do.

Reply:
Done.